contracts/src/Teleporter/ReceiptQueue.sol

```diff
@@ -8,90 +8,97 @@ pragma solidity 0.8.18;
+    }
 
-    /**
-     * @dev Emitted when a receipt is taken off the front of the queue.
-     */
-    event Dequeue(
-        uint256 indexed messageID,
-        address indexed relayerRewardAddress
-    );
+    // The maximum number of receipts to include in a single message.
+    uint256 private constant _MAXIMUM_RECEIPT_COUNT = 5;
 
     // Errors
-    error Unauthorized();
     error EmptyQueue();
-
-    constructor() {
-        owner = msg.sender;
-    }
+    error OutofIndex();
 
     /**
      * @dev Adds a receipt to the queue.
-     *
-     * Requirements:
-     *
-     * - `msg.sender` must be the owner.
      */
-    function enqueue(TeleporterMessageReceipt calldata receipt) external {
-        if (msg.sender != owner) {
-            revert Unauthorized();
-        }
-
-        queue[last++] = receipt;
-
-        emit Enqueue(receipt.receivedMessageID, receipt.relayerRewardAddress);
+    function enqueue(
+        TeleporterMessageReceiptQueue storage queue,
+        TeleporterMessageReceipt memory receipt
+    ) internal {
+        queue.data[queue.last++] = receipt;
     }
 
     /**
      * @dev Removes the oldest open receipt from the queue.
      *
      * Requirements:
-     *
-     * - `msg.sender` must be the owner.
      * - The queue must be non-empty.
      */
-    function dequeue()
-        external
+    function dequeue(
+        TeleporterMessageReceiptQueue storage queue
+    )
+        internal
         returns (TeleporterMessageReceipt memory result)
     {
-        if (msg.sender != owner) {
-            revert Unauthorized();
-        }
-
-        uint256 first_ = first;
+        uint256 first_ = queue.first;
+        if (queue.last == first_) revert EmptyQueue();
+        result = queue.data[first_];
+        delete queue.data[first_];
+        queue.first = first_ + 1;
+    }
 
-        if (last == first_) {
-            revert EmptyQueue(); // empty queue
+    /**
+     * @dev Returns the outstanding receipts for the given chain ID that should be included in the next message sent.
+     */
+    function getOutstandingReceiptsToSend(
+        TeleporterMessageReceiptQueue storage queue
+    )
+        internal
+        returns (TeleporterMessageReceipt[] memory result)
+    {
+        // Get the current outstanding receipts for the given chain ID.
+        // If the queue contract doesn't exist, there are no outstanding receipts to send.
+        uint256 outstandingReceiptsSize = size(queue);
+        if (outstandingReceiptsSize == 0) {
+            return new TeleporterMessageReceipt[](0);
         }
 
-        result = queue[first_];
-
-        delete queue[first_];
-        first = first_ + 1;
+        // Calculate the result size as the minimum of the number of receipts and maximum batch size.
+        uint256 resultSize = _MAXIMUM_RECEIPT_COUNT;
+        if (outstandingReceiptsSize < _MAXIMUM_RECEIPT_COUNT) {
+            resultSize = outstandingReceiptsSize;
+        }
```

Comment:
I think this would save a bit of gas by not declaring an extra variable. We'd probably want to rename `outstandingReceiptsSize` to `resultSize` though.
```suggestion
        if (outstandingReceiptsSize > _MAXIMUM_RECEIPT_COUNT) {
            outstandingReceiptsSize = _MAXIMUM_RECEIPT_COUNT;
        }
```

Reply:
fixed